examples/examples/jsonrpsee_as_service.rs

```diff
@@ -70,26 +75,97 @@ struct AuthorizationMiddleware<S> {
 	transport_label: &'static str,
 }
 
-impl<'a, S> RpcServiceT<'a> for AuthorizationMiddleware<S>
+impl<S> AuthorizationMiddleware<S> {
+	/// Authorize the request by checking the `Authorization` header.
+	///
+	///
+	/// In this example for simplicity, the authorization value is not checked
+	// and used because it's just a toy example.
+	fn auth_method_call(&self, req: &Request<'_>) -> bool {
+		if req.method_name() == "trusted_call" {
+			let Some(Ok(_)) = self.headers.get(AUTHORIZATION).map(|auth| auth.to_str()) else { return false };
+		}
+
+		true
+	}
+
+	/// Authorize the notification by checking the `Authorization` header.
+	///
+	/// Because notifications are not expected to return a response, we
+	/// return a `MethodResponse` by injecting an error into the extensions
+	/// which could be read by other middleware or the server.
+	fn auth_notif(&self, notif: &Notification<'_>) -> bool {
+		if notif.method_name() == "trusted_call" {
+			let Some(Ok(_)) = self.headers.get(AUTHORIZATION).map(|auth| auth.to_str()) else { return false };
+		}
+
+		true
+	}
+}
+
+impl<S> RpcServiceT for AuthorizationMiddleware<S>
 where
-	S: Send + Clone + Sync + RpcServiceT<'a>,
+	S: Send + Clone + Sync + RpcServiceT<Response = MethodResponse, Error = Infallible> + 'static,
+	S::Error: Into<BoxError> + Send + 'static,
+	S::Response: Send,
 {
-	type Future = ResponseFuture<S::Future>;
+	type Error = S::Error;
+	type Response = S::Response;
+
+	fn call<'a>(&self, req: Request<'a>) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + 'a {
+		let this = self.clone();
+		let auth_ok = this.auth_method_call(&req);
+
+		async move {
+			// If the authorization header is missing, it's recommended to
+			// to return the response as MethodResponse::error instead of
+			// returning an error from the service.
+			//
+			// This way the error is returned as a JSON-RPC error
+			if !auth_ok {
+				return Ok(MethodResponse::error(req.id, auth_reject_error()));
+			}
+			this.inner.call(req).await
+		}
+	}
 
-	fn call(&self, req: Request<'a>) -> Self::Future {
-		if req.method_name() == "trusted_call" {
-			let Some(Ok(_)) = self.headers.get(AUTHORIZATION).map(|auth| auth.to_str()) else {
-				let rp = MethodResponse::error(req.id, ErrorObject::borrowed(-32000, "Authorization failed", None));
-				return ResponseFuture::ready(rp);
-			};
+	fn batch<'a>(&self, batch: Batch<'a>) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + 'a {
+		// Check the authorization header for each entry in the batch.
+		let entries: Vec<_> = batch
+			.into_iter()
+			.filter_map(|entry| match entry {
+				Ok(BatchEntry::Call(req)) => {
+					if self.auth_method_call(&req) {
+						Some(Ok(BatchEntry::Call(req)))
+					} else {
+						// If the authorization header is missing, we return
+						// a JSON-RPC error instead of an error from the service.
+						Some(Err(ErrorResponse::new(req.id, auth_reject_error())))
```

Comment:
How does `Err(ErrorResponse::new(req.id, auth_reject_error()))` differ from `Ok(MethodResponse::error(req.id, auth_reject_error()))`? Do they both lead to the same behaviour (ie a JSONRPC error being returned for that call)?

Reply:
This must be outdated code but returning an error from the RpcServiceT would lead to that we would actually catch in the server/RpcService impl and convert into a JSON-RPC error which is annoying.

I have restricted this now such that all RpcServiceT must use Error = Infallible, it much better than that we provide an error message from the Debug/String impl from the Error type.

Reply:
Ah I see it's BatchEntry and BatchEntryErr; that works for me and feels clearer :)